ga4gh/server/static/main.js

```diff
@@ -2,50 +2,97 @@
-    xhr.setRequestHeader('Authorization', 'Bearer ' + session_id);
-    xhr.send(JSON.stringify({}));
-    xhr.onload = function() {
-
-        if (xhr.status != 200) {
+    makeRequest("datasets/search", {}).then(function(response) {
+
+        const data = JSON.parse(response);
+        const listOfDatasetId = data['results']['datasets'];
+
+        let dropdown = document.getElementById("dropdown-menu");
+
+        for (let i = 0; i < listOfDatasetId.length; i++) {
+            if (!finalDatasetId.includes(listOfDatasetId[i]['id'])) {
+                finalDatasetId.push(listOfDatasetId[i]['id']);
+                finalDatasetName.push(listOfDatasetId[i]['name']);
+            }
+        }
+
+        for (let j = 0; j < finalDatasetId.length; j++) {
+            dropdown.innerHTML += '<a class="dropdown-item" id="refresh" href="javascript:void(0)" onclick="refreshDataset(' + j + ')">' + finalDatasetName[j] + '</a>'
+        }
+
+        datasetId = finalDatasetId[0];
+        $('#dropdownMenuLink').html('<i class="fas fa-database"></i> ' + finalDatasetName[0]);
+
+        if (location.hash == "#candig_patients" || location.hash == "#gene_search" || location.hash == "#sample_analysis") {
+            $('.nav-tabs a[href="' + location.hash + '"]').tab('show');
+        }
+
+        else $('.nav-tabs a[href="#candig"]').tab('show');
+
+    }, function(Error) {
             document.getElementById('tab-content').style.display = "none";
             alertBuilder("No data currently available. Please contact a system administrator for assistance.")
-        } else {
-            const data = JSON.parse(this.responseText);
-            const listOfDatasetId = data['results']['datasets'];
+    })
+});
 
-            if (listOfDatasetId.length == 0) {
-                alertBuilder("Sorry, but it seems like no data is available at the moment..");
-            } else {
-                let dropdown = document.getElementById("dropdown-menu");
+function makeRequest(path, body) {
+    return new Promise(function(resolve, reject) {
+    let results = []
+    let key;
 
-                for (let i = 0; i < listOfDatasetId.length; i++) {
-                    if (!finalDatasetId.includes(listOfDatasetId[i]['id'])) {
-                        finalDatasetId.push(listOfDatasetId[i]['id']);
-                        finalDatasetName.push(listOfDatasetId[i]['name']);
-                    }
-                }
+    // Initialize the request with empty pageToken
+    return repeatRequest("");
+
+    function repeatRequest(pageToken){
+        body["pageToken"] = pageToken;
+
+        var xhr = new XMLHttpRequest();
+        xhr.open('POST', prepend_path + path, true);
+        xhr.setRequestHeader('Content-Type', 'application/json');
+        xhr.setRequestHeader('Accept', 'application/json');
+        xhr.setRequestHeader('Authorization', 'Bearer ' + session_id);
+        xhr.onload = function() {
+            if (xhr.status === 200) {
 
```

Comment:
maybe redundant check (?)

Reply:
it seems like `onload` in equivalent to `onreadystate == 4`, so it should not be a redundant check here.